src/PJ_cart.c

```diff
@@ -0,0 +1,253 @@
+ *
+ ******************************************************************************
+ * Copyright (c) 2016, Thomas Knudsen / SDFE
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *****************************************************************************/
+
+#define PJ_LIB__
+#include <proj.h>
+#include <projects.h>
+#include <assert.h>
+#include <stddef.h>
+#include <math.h>
+#include <errno.h>
+PROJ_HEAD(cart,    "Geodetic/cartesian conversions");
+
+
+/**************************************************************
+                CARTESIAN / GEODETIC CONVERSIONS
+***************************************************************
+    This material follows:
+
+    Bernhard Hofmann-Wellenhof & Helmut Moritz:
+    Physical Geodesy, 2nd edition.
+    Springer, 2005.
+
+    chapter 5.6: Coordinate transformations
+    (HM, below),
+
+    and
+
+    Wikipedia: Geographic Coordinate Conversion,
+    https://en.m.wikipedia.org/wiki/Geographic_coordinate_conversion
+
+    (WP, below).
+
+    The cartesian-to-geodetic conversion is based on Bowring's
+    celebrated method:
+
+    B. R. Bowring:
+    Transformation from spatial to geographical coordinates
+    Survey Review 23(181), pp. 323-327, 1976
+
+    (BB, below),
+
+    but could probably use some TLC from a newer and faster
+    algorithm:
+
+    Toshio Fukushima:
+    Transformation from Cartesian to Geodetic Coordinates
+    Accelerated by Halley’s Method
+    Journal of Geodesy, February 2006
+
+    (TF, below).
+
+
+    These routines are probably not as robust at those in
+    geocent.c, at least thay haven't been through as heavy
+    use as their geocent sisters. Some care has been taken
+    to avoid singularities, but extreme cases (e.g. setting
+    es, the squared eccentricity, to 1), will cause havoc.
+
+**************************************************************/
+
+
+static void freeup (PJ *P) {
+    pj_freeup_plain (P);
```

Comment:
It is defined in pj_malloc.c and has a slightly different prototype than pj_dealloc: It takes a PJ as arg and frees its opaque object. It can be used to simplify/eliminate the freeup code for a large number of projections.